src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java

```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+        if (jmsSelector.equals("_PREV_SAMPLER_")) {
+            if (sampler instanceof JMSSampler) {
+                jmsSelector = sr.getResponseMessage();
+            }
+        }
+        int sampleCounter = 0;
+        int sampleTries = 0;
+        String result = null;
+
+        StringBuilder buffer = new StringBuilder();
+        StringBuilder propBuffer = new StringBuilder();
+
+        do {
+            result = browseQueueForConsumption(sendQueue, jmsSelector, res, buffer, propBuffer);
+            if (result != null) {
+                sb.append(result);
+                sb.append('\n');
+                sampleCounter++;
+            }
+            sampleTries++;
+        } while ((result != null) && (sampleTries < getNumberOfSamplesToAggregateAsInt()));
+
+        res.setResponseMessage(sampleCounter + " samples messages received");
+        res.setResponseData(buffer.toString().getBytes()); // TODO - charset?
+        res.setResponseHeaders(propBuffer.toString());
+        if (sampleCounter == 0) {
+            res.setResponseCode("404");
+            res.setSuccessful(false);
+        } else {
+            res.setResponseCodeOK();
+            res.setSuccessful(true);
+        }
+        res.setResponseMessage(sampleCounter + " message(s) received successfully");
+        res.setSamplerData(getNumberOfSamplesToAggregateAsInt() + " messages expected");
+        res.setSampleCount(sampleCounter);
+    }
+
+    private void handleRequestResponse(SampleResult res) throws JMSException {
+        TextMessage msg = createMessage();
+        if (!useTemporyQueue()) {
+            LOGGER.debug("NO TEMP QUEUE");
+            msg.setJMSReplyTo(receiveQueue);
+        }
+        LOGGER.debug("Create temp message");
+        Message replyMsg = executor.sendAndReceive(msg,
+                isNonPersistent() ? DeliveryMode.NON_PERSISTENT : DeliveryMode.PERSISTENT,
+                Integer.parseInt(getPriority()), Long.parseLong(getExpiration()));
+        res.setRequestHeaders(Utils.messageProperties(msg));
+        if (replyMsg == null) {
+            res.setResponseMessage("No reply message received");
+        } else {
+            if (replyMsg instanceof TextMessage) {
+                res.setResponseData(((TextMessage) replyMsg).getText(), null);
+            } else {
+                res.setResponseData(replyMsg.toString(), null);
+            }
+            res.setResponseHeaders(Utils.messageProperties(replyMsg));
+            res.setResponseOK();
+        }
+    }
+
+    private String browseQueueForConsumption(Queue queue, String jmsSelector, SampleResult res, StringBuilder buffer,
+            StringBuilder propBuffer) {
+        String retVal = null;
+        try {
+            QueueReceiver consumer = session.createReceiver(queue, jmsSelector);
+            Message reply = consumer.receive(Long.valueOf(getTimeout()));
+            LOGGER.debug("Message: " + reply);
+            consumer.close();
```

Comment:
You should use try with resource pattern to ensure close happens ? Or use try/finally and closeQuietly depending on what you want to do